### PR TITLE
perf(secrets): run the redactor's whole-payload scans only where a match can begin

### PR DESCRIPTION
**Pull request by alexander-turner**
Claimed area: `secrets/{engine,prefilter,invisible}.py` — the redactor's three remaining whole-payload passes. No overlap with any other open PR.

## Summary

Every one of the redactor's four whole-payload passes swept the entire text even when it could prove, cheaply, that almost none of the text could contain a match. This PR replaces each sweep with a scan bounded by what the pattern itself admits, and each per-character Python loop with a C scan.

| payload | before | after |
| --- | --- | --- |
| 8.5 MB mixed corpus | 10.08 s | 5.20 s |
| 8.5 MB, no invisibles or case exotics | 10.03 s | 4.58 s |
| 2 MB | 2.30 s | 0.90 s |
| single short line | 21.2 ms | 14.9 ms |

**The <1s target holds to about 2 MB, not to 8.5 MB.** 8.5 MB is still ~5 s, and the residual is now dominated by `detect-secrets`' own `scan_line` on the candidates that survive, not by the sweeps this PR narrowed. No measured case is slower than before.

Nothing here changes *what* counts as a secret. `scan_line` remains the sole detection oracle; the new code only decides *where* it is asked. Every narrowing is a sound necessary condition derived from the detector's own `denylist`, so the failure direction is extra work, never a missed key.

## Changes

- **`d5cebc2` — fold by `str.replace` chain, not `str.translate`.** `translate` over a dict pays a hash lookup per character of the whole payload; five `replace` calls are C substring scans that return the input untouched when their character is absent. An import-time `_SELF_FEEDING` guard raises if a future fixup entry would make the chain order-dependent (which is the one thing that would break equivalence with the single `translate`).
- **`2ff39dd` — decide the invisible strip with a range-collapsed character class.** The presence probe is now a C scan that stops at the first hit, and the strip splices the survivors between the located deletions instead of re-deriving them with `translate`.
- **`a2ab7e5` — three engine changes on one pass over `_redact_core`:**
  - `_redact_field_values` tries `FIELD_VALUE_RE` only at the positions a new flagless `_FIELD_NAME_ANCHOR` matches, rather than at all of them. Two import-time guards hold the soundness claim: one that the regex still *opens* with the credential-noun alternation, one that the vocabulary is lowercase (the anchor is matched against folded text). The anchor is deliberately not `re.IGNORECASE` — CPython disables its literal-prefix optimization outright for an IGNORECASE pattern, worth 1.8 s vs 0.17 s over 8 MB here.
  - The cross-line sweep is narrowed to `LiteralProbe.plan`'s windows. Any pattern no window can bound still sweeps the whole text individually, gated on its own literals.
  - `_redact_line` accepts `invisible_free`, so a payload proven invisible-free once does not pay a strip per line.
- **`24cd6ae` — `str.isascii()` shortcut and cached AST parses.** `isascii` is O(1) in CPython and every charset member is above ASCII (derived, not assumed), so it is a free proof of invisible-freeness — and on clean ASCII it beats the class scan. `inspect_width` and `required_literals` are `functools.cache`d, since adding a second AST parse per pattern had made small payloads slower than baseline.
- **`85d8a42` — a word boundary stays readable inside its window.** `endpos` makes a window's right edge look like the end of the string, which *flips* `\b`/`\B` there and drops a real match with no error (`re.compile(r'a-\b').search('a-b', 0, 2)` is `None` where the full search matches). `_inspect_width` now costs a boundary assertion one character of extent, and an assertion missing from `_AT_INSPECT_WIDTHS` makes the pattern unwindowable rather than wrongly bounded. Found in review; no live denylist carries one today, so nothing was leaking.
- **`d1719a4` — an empty invisible charset strips nothing instead of raising.** A memberless class spells `[]`, which `re` rejects as a syntax error, so an explicit empty override blew up as soon as a payload was non-ASCII enough to get past the `isascii` shortcut. It compiles to a never-matching class now, which also makes `invisible_run_pattern` match exactly the empty string. Also hardens the per-line-strip count (see below).
- **`966908e`, `b78fac2`, `0486360`, `75063e7`** — the sweep branch spelled as a plain `if`/`else`, the `invisible_free=` keyword at its call site, tests, and docstrings brought under the 20-line cap.

## Review focus

Read **`prefilter.py::LiteralProbe.plan`** first — it is the only new claim that can lose a secret if wrong. Its contract: for each occurrence of literal `l` at `[start, end)`, the window `(end - w, start + w)` contains every match of any pattern that literal indexes, where `w` is `inspect_width`'s upper bound on how far right of its start that pattern can *read*.

Three invariants to check together, since they span files:

1. **`inspect_width` bounds the *inspected* extent, not the consumed one** — a trailing lookahead's width is added in, and so is the one character a `\b`/`\B` reads to its right, because `endpos` truncates the text and a truncated assertion turns a real match into a miss. Any opcode (or assertion) it does not enumerate returns `None`, which forces a full scan.
2. **Ranges are for `finditer(text, pos, endpos)`, never for slicing.** `pos` leaves the text to its left readable, so a lookbehind, `^` under `MULTILINE`, and `$` see the same text a full sweep shows them. Slicing would silently fabricate a start-of-string.
3. **`plan().windows is None` means "could not narrow"; empty tuple means "matches nowhere".** Conflating those two is exactly how a caller skips a scan the probe never ruled out — an earlier draft did, and the cross-line stage degenerated to one `scan_line` per character.

The part I am least sure of: `_coalesce` merges **abutting** ranges as well as overlapping ones, because two adjacently-scanned ranges would let a match straddling their shared edge be found by neither. I believe that is the only edge case of its kind, but it is the sort of off-by-one that a test corpus can miss.

Also worth a second pair of eyes: `_redact_field_values`' `at` cursor. `Pattern.search(s, pos)` **clamps** a `pos` past the end back to `len(s)` and returns a match there rather than `None`, so a zero-width-capable anchor is re-found forever without the strictly-increasing cursor. That cost three killed test runs before it was isolated; there is a regression test for it.

## Tests / verification

- `uv run --extra dev pytest tests/` — **3506 passed, 5 skipped**; `ruff format`/`check` clean; `plugin/dist/redactor/` zipapps rebuilt so the `committed == regenerate(source)` freshness check holds.
- **Byte-identity, not sampling.** `_redact_field_values` vs `FIELD_VALUE_RE.sub` over 12 corpora (including this engine's own source, overlapping credential nouns, dotted capital I, long s, and multiline operators): 0 mismatches. The windowed cross-line sweep vs the unwindowed one over the 8.5 MB corpus plus 7 adversarial payloads: 0 mismatches.
- **Every new guard is red before its fix.** `test_windows_keep_a_word_boundary_reading_real_bytes` and `test_an_empty_charset_strips_nothing_rather_than_failing_to_compile` were both run against the pre-fix source and both fail there (dropped match; `re.error: unterminated character set`).
- **Non-vacuity for the strip skip, measured.** The per-line-strip count is now driven by a payload that is non-ASCII *and* carries a credential literal per line — either property missing and the count is 0 either way, since the `isascii` shortcut or the line probe answers first. With the skip: 2 scans. With `invisible_free` forced False: 402. Both directions are pinned as tests. `test_skipping_the_strip_on_a_dirty_payload_would_leak_the_key` separately forces the skip on a payload that *does* carry a zero-width character and asserts the key survives.
- New unit coverage: `inspect_width` per-opcode and per-assertion (including every `None` case — `a+`, `a*`, `a{2,}`, `a+?`, a backreference), `_coalesce`, window soundness (`every match of a windowable pattern lies in a window`), the field-value anchor matching wherever the regex can start, the fold-refused fallback, deadline propagation through both new loops, and the zero-width-anchor termination guard.
- Negative side: the existing legitimate-input corpora assert zero findings unchanged, which is the property that would break if any of these narrowings had turned into a *widening*.

## Decisions made

- **No width cap on `inspect_width`.** Capping it at ~1024 would take the cross-line stage from 763 ms to 349 ms, by pushing the `eyJ…{1,65536}` JWT pattern onto a full sweep (its distinctive literal prefix makes `re` skip cheaply there). That needs an arbitrary constant, so the rule shipped is constant-free instead: window a literal iff its own coalesced windows cover less than the whole payload. Under the alternative, 8.5 MB drops by roughly 400 ms and the code grows a magic number to justify and tune.
- **A boundary assertion costs one character of extent rather than making the pattern unwindowable.** The review suggested refusing the bound outright for `\B`. Refusing is also sound, but it is both narrower than the bug (`\b` flips under truncation too, for a pattern ending in a non-word character) and lossier: an upstream detect-secrets upgrade that adds a `\b` to a denylist would silently cost a full sweep. Costing it 1 keeps the pattern windowable and fixes both spellings.
- **`a2ab7e5` carries three engine changes rather than three commits.** They were verified together as one diff over `_redact_core`, and splitting a byte-identity-verified diff after the fact risked mangling it. Its subject names all three.
- **`strip_invisible` is now a one-line delegation to `strip_invisible_with_map`.** That is a pass-through wrapper the style rules normally reject, but it is the module's public, independently-tested boundary and its name documents the contract (deletion-only normalization, no offset map), so it is kept rather than inlined at ~20 call sites.

## Checklist

- [x] Commits follow [Conventional Commits](https://www.conventionalcommits.org/); each subject reads as a user-facing release note.
- [x] Tests and lint pass locally (`pytest tests/`, `ruff format`, `ruff check`).
- [x] New or changed detectors have negative tests (zero findings on legitimate input) and pin their invariants with property tests. No detector changed here — the existing negative corpora are what prove the narrowings did not widen.
- [x] No real secrets/tokens in the diff, tests, or description.
- [x] `CHANGELOG.md` and `package.json` version left untouched.

### CODE REVIEW

File: python/agent_sanitizer/secrets/prefilter.py [hunk 1]
```diff
@@ -183,10 +216,81 @@ def _required(seq) -> frozenset[str] | None:
...
+    this does not understand costs a full scan instead of a wrong bound — a
+    backreference is exactly that case: its width is another group's, which is
+    not readable from here.
+    """
+    total = 0
+    for op, av in seq:
+        if op in (_c.LITERAL, _c.NOT_LITERAL, _c.IN, _c.ANY):
+            total += 1
+        elif op is _c.AT:
+            continue  # A zero-width anchor consumes and inspects nothing.
```
[the remark this answers is not in the document]
Comment (alexander-turner) [reply python/agent_sanitizer/secrets/prefilter.py]: Real, and fixed in 85d8a42 — but wider than the suggestion, because `\b` is unsound here too, not just `\B`:

```python
>>> re.compile(r'a-\b').search('a-b')        # <re.Match span=(0, 2)>
>>> re.compile(r'a-\b').search('a-b', 0, 2)  # None
```

`\b` at position `p` is `isword(s[p-1]) != isword(s[p])`. Truncation forces `isword(s[p])` to False, which *flips* the answer whenever `s[p-1]` is a non-word character and `s[p]` is a word character — so a pattern ending `-\b` loses its match exactly like `abc\B` does. Refusing only the three `NON_BOUNDARY` spellings would have left that hole open.

The fix is a table of how far right each zero-consuming assertion *reads*, consumed by `_inspect_width`:

- `\b`/`\B` (all six spellings, incl. LOC/UNI) → **1**, so the window reaches one character past the assertion and it reads a real byte instead of a fabricated end-of-string.
- start-of-text and end-of-text anchors → **0**. `^` is unaffected because `pos` does not fabricate a start; `$`/`\Z`/`$`-under-MULTILINE only become *more* true under truncation, which costs a candidate rather than losing one.
- anything absent from the table → `None`, i.e. unwindowable. That keeps the "unrecognized construct must widen, never narrow" rule you cite, and makes it the default for any future opcode rather than something to remember.

Costing a boundary 1 rather than refusing the bound also keeps such a pattern windowable, so an upstream detect-secrets upgrade that adds a `\b` does not silently cost a full sweep either.

Covered by `test_windows_keep_a_word_boundary_reading_real_bytes` (red before the fix: the match at 400 is dropped) plus `(r"abc\b", 4)`, `(r"abc\B", 4)`, `(r"\babc", 4)`, `(r"abc\Z", 3)` beside the existing `^abc$` row.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/secrets/test_secrets_invisible.py [hunk 2]
```diff
@@ -207,3 +211,82 @@ def _boom():
...
+    return result, counter.searches
+
+
+def test_a_clean_payload_pays_no_per_line_strip(monkeypatch):
+    """The whole-payload strip already proves no line holds an invisible
+    character, so the per-line strip is skipped outright — one scan for the
+    payload, not one per line."""
+    lines = [f"line {i} nothing to see" for i in range(400)]
+    _, searches = _count_strips(monkeypatch, "\n".join(lines))
+    assert searches < len(lines)
```
[the remark this answers is not in the document]
Comment (alexander-turner) [reply tests/secrets/test_secrets_invisible.py]: Correct — the test was vacuous, fixed in d1719a4. One thing to flag, because the suggested payload does **not** fix it: I measured it, and `f"line {i} nothing to sée"` still counts **2 with the skip and 2 with `invisible_free` forced False**.

Non-ASCII gets the payload past the `str.isascii` shortcut, but there is a second gate: those lines carry no credential literal, so `_line_probe().candidates` rules every one of them out and `_redact_line` is never called for them at all. Nothing reaches the per-line strip either way.

The payload therefore needs *both* properties, and the comment on it now says so: non-ASCII, **and** a credential literal per line so the line probe actually hands the line over. `akia note {i} nothing to sée` does that and is not key-shaped, so nothing is redacted and the strip count is the only thing that differs:

| | searches |
| --- | --- |
| with the skip | 2 |
| `invisible_free` forced False | 402 |

Bound tightened to `< 10` as suggested, and the reverted case is now pinned as its own test (`test_without_the_skip_every_line_pays_its_own_strip`, asserting `> 400`) so the count cannot quietly go vacuous again.

---
_Generated by [Claude Code](https://claude.ai/code)_
